src/mpi4py/util/pkl5.py

```diff
@@ -0,0 +1,534 @@
+# Author:  Lisandro Dalcin
+# Contact: dalcinl@gmail.com
+"""Pickle-based communication using protocol 5."""
+
+import sys as _sys
+import struct as _struct
+
+from .. import MPI
+from ..MPI import (
+    PROC_NULL,
+    ANY_SOURCE,
+    ANY_TAG,
+    Status,
+)
+
+from ..MPI import (
+    _typedict,
+    _comm_lock,
+    memory as _memory,
+    Pickle as _Pickle,
+)
+
+if _sys.version_info >= (3, 8):
+    from pickle import (
+        dumps as _dumps,
+        loads as _loads,
+        HIGHEST_PROTOCOL as _PROTOCOL,
+    )
+else:  # pragma: no cover
+    try:
+        from pickle5 import (
+            dumps as _dumps,
+            loads as _loads,
+            HIGHEST_PROTOCOL as _PROTOCOL,
+        )
+    except ImportError:
+        _PROTOCOL = MPI.Pickle().PROTOCOL
+
+        def _dumps(obj, *_p, **_kw):
+            return MPI.pickle.dumps(obj)
+
+        def _loads(buf, *_p, **_kw):
+            return MPI.pickle.loads(buf)
+
+
+def _buffer_handler(protocol, threshold):
+    bufs = []
+    if protocol is None or protocol < 0:
+        protocol = _PROTOCOL
+    if protocol < 5:
+        return bufs, None
+    buffer_len = len
+    buffer_raw = _memory
+    buffer_add = bufs.append
+    def buf_cb(buf):
+        buf = buffer_raw(buf)
+        if buffer_len(buf) >= threshold:
+            buffer_add(buf)
+            return False
+        return True
+    return bufs, buf_cb
+
+
+class Pickle(_Pickle):
+    """Pickle/unpickle Python objects using out-of-band buffers."""
+
+    THRESHOLD = 1024**2 // 4  # 0.25 MiB
```

Comment:
I don't understand your question. If pickle 5 is not available, `THRESHOLD` is irrelevant (not used), and everything is pickled in-band. In other words, if pickle 5 is not available, then the `dumps()` method returns `data, bufs` with all of the pickled data in `data` an an empty `bufs` list.

Reply:
Ah, I should have been more explicit, sorry. I meant why did we pick 0.25 MiB here and nothing some larger (or smaller) values?

Reply:
Oh, sorry, I totally misunderstood your question. No, this is not a limitation of pickle 5. If you use out-of-band buffers, you send/receive more messages. If messages are small, latency dominates. So, in my admittedly very limited testing on my workstation, the 0.25 MiB threshold was the value where out-of-band communication started to pay off in terms of wall-clock time. My test used a shared memory channel, which is very low latency. Perhaps we should do more exhaustive testing in the future. Anyway, this should not delay the merge of this MR, if needed, we just need to update this single line with a better default value.

Reply:
Thanks for clarifying. I am just curious about this pick and have no intention to block this PR because of this single line 🙂 I think if there's a need we can revisit later, possible allowing it to be controlled by an env var?